### PR TITLE
Tests local storage and ensures nv shared compatibility

### PR DESCRIPTION
**Pull request by NoraHagmeyer**
## Description
For nv shared and cooperative dimensions, we enforce that the iteration count is a provably a multiple of the block_size. This ensures that we don't run into dead lock issues but it is a very conservative treatment.

To improve that, we either
a) have to be able to pull the barrier out of the conditional code
b) add (iteration) padding to avoid conditional code altogether (basically what I did by hand in 2016, when cuda was not too good yet)

The biggest junk is changing tests to use numeric bounds to check that all other functionality satyed the same, and adding additional unit test for symbolic sizes and to ensure rejection where we now expect it.

## Related Issues

- Closes #

## Type of Change
- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Refactor / internal change (no functional change)
- [ ] Documentation update

## Quality Checklist
*Please ensure the following are completed before requesting review:*

* [x] Runtime-compatible (no externally visible existing method has its parameters changed, was deleted or has its exception behavior changed)
* [x] Compile-time compatible (new arguments & properties added have default values, but users need to recompile against the new version to work)

IR:
* [ ] Introduces new types into the SDFG IR
* [ ]  Changed / added properties on SDFG IR Elements (LibraryNodes, ControlFlowNodes DataFlowNodes etc.)
  *  [ ] New / modified types have serialization / deserialization support
  *  [ ] Deserialization is backward compatible (new serialization can deserialize old format and represent it in the new format if applicable)
  *  [ ] Serialized format is backward compatible (previous deserializer can read new serialized format without incorrectness (just lacking additional guarantees that are optional or did not exist before)
  *  [ ] Serializer has an option to emit the previous format

### Unit Tests
- [x] New unit tests have been added for the changes.
- [x] Existing unit tests pass locally.
- [x] Edge cases and error paths are covered.

### Integration Tests
- [ ] New integration tests have been added (or existing ones updated).
- [ ] Integration tests pass locally.
- [ ] Interactions with other components/services are verified.

### Documentation
- [ ] Public APIs, options, and behavior changes are documented.
- [ ] README / docs site / inline comments updated where applicable.
- [ ] Changelog / release notes updated (if applicable).

### Test Coverage
- [ ] Test coverage has not decreased as a result of this change.
- [ ] Coverage report reviewed and acceptable.
- [ ] Critical / new code paths are covered.

## How Has This Been Tested?
Describe the tests you ran to verify your changes and provide instructions to reproduce.
Include relevant details for your test configuration (OS, Python/LLVM/Docc version, hardware/target).

## Additional Context
Add any other context about the PR here (design decisions, trade-offs, follow-ups).

### CODE REVIEW

File: opt/src/transformations/in_local_storage.cpp [hunk 1]
```diff
@@ -138,39 +154,113 @@ bool InLocalStorage::can_be_applied(builder::StructuredSDFGBuilder& builder, ana
...
+        // A bare symbolic bound (e.g. `N`) is not syntactically a multiple of
+        // the block size, but an assumption established earlier in the program
+        // may pin it to one. We fold such equality facts into the trip count
+        // before the divisibility test: a symbol is replaced by an expression
+        // it is provably equal to when its tight lower/upper bounds coincide,
+        // or when the same expression appears as both a lower and an upper
+        // bound. This lets `N` be accepted when an upstream assumption
+        // established `N == 32*P` or `N == 1024` (symbol / condition
+        // propagation), while a genuinely unconstrained `N` is still rejected.
+        auto& assumptions_analysis = analysis_manager.get<analysis::AssumptionsAnalysis>();
```
Comment (NoraHagmeyer) [inline opt/src/transformations/in_local_storage.cpp:196]: @lukastruemper Can you have a look here and see if we can do something smarter right out of the box?

File: opt/src/transformations/in_local_storage.cpp [hunk 2]
```diff
@@ -138,39 +154,113 @@ bool InLocalStorage::can_be_applied(builder::StructuredSDFGBuilder& builder, ana
...
+            auto condition = ancestor_map->condition();
+            if (!SymEngine::is_a<SymEngine::StrictLessThan>(*condition)) {
+                // Cannot reason about the trip count → conservatively reject.
+                return false;
+            }
+            auto stl = SymEngine::rcp_static_cast<const SymEngine::StrictLessThan>(condition);
+            auto rhs = stl->get_args()[1];
+            auto iter_count = symbolic::sub(rhs, ancestor_map->init());
+            iter_count = resolve_with_assumptions(iter_count, assumptions_analysis.get(*ancestor_map, true));
+            if (!provably_multiple_of_block(iter_count, block_size)) {
```
Comment (NoraHagmeyer) [inline opt/src/transformations/in_local_storage.cpp:256]: Can be removed if we have a safe way to treat conditional code and barriers
